### ISSUE TITLE
Bundled gems with extensions fail to install during build on Windows

### ISSUE THREAD

**Issue by headius** [jruby/jruby#6174]
Specifically, the io-console gem, which has a no-op extconf on JRuby, fails to install with the following error:

```

operable program or batch file.
[INFO] JRuby Integration Tests ............................ SKIPPED

[INFO] ------------------------------------------------------------------------
extconf failed, exit code 1
[INFO] BUILD FAILURE

[INFO] ------------------------------------------------------------------------
Gem files will remain installed in D:/a/1/s/lib/target/rubygems/gems/io-console-0.5.6 for inspection.
[INFO] Total time:  04:39 min
Results logged to D:/a/1/s/lib/target/rubygems/extensions/universal-java-1.8/2.3.0/io-console-0.5.6/gem_make.out
[INFO] Finished at: 2020-04-13T20:42:05Z

[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal io.takari.polyglot:polyglot-maven-plugin:0.3.1:execute (install_gems) on project jruby-stdlib: (BuildError) ERROR: Failed to build gem native extension.
[ERROR] 
[ERROR]     current directory: D:/a/1/s/lib/target/rubygems/gems/io-console-0.5.6/ext/io/console
[ERROR] java -cp ;C:\Users\VssAdministrator\.m2\repository\io\takari\polyglot\polyglot-ruby\0.3.1.1\polyglot-ruby-0.3.1.1.jar;C:\Users\VssAdministrator\.m2\repository\io\takari\polyglot\polyglot-common\0.3.1\polyglot-common-0.3.1.jar;C:\Users\VssAdministrator\.m2\repository\org\jruby\jruby\9.1.17.0\jruby-9.1.17.0.pom;C:\Users\VssAdministrator\.m2\repository\org\jruby\jruby-core\9.1.17.0\jruby-core-9.1.17.0.jar;C:\Users\VssAdministrator\.m2\repository\com\github\jnr\jnr-netdb\1.1.6\jnr-netdb-1.1.6.jar;C:\Users\VssAdministrator\.m2\repository\com\github\jnr\jnr-enxio\0.16\jnr-enxio-0.16.jar;C:\Users\VssAdministrator\.m2\repository\com\github\jnr\jnr-x86asm\1.0.2\jnr-x86asm-1.0.2.jar;C:\Users\VssAdministrator\.m2\repository\com\github\jnr\jnr-unixsocket\0.17\jnr-unixsocket-0.17.jar;C:\Users\VssAdministrator\.m2\repository\com\github\jnr\jnr-posix\3.0.44\jnr-posix-3.0.44.jar;C:\Users\VssAdministrator\.m2\repository\com\github\jnr\jnr-constants\0.9.9\jnr-constants-0.9.9.jar;C:\Users\VssAdministrator\.m2\repository\com\github\jnr\jffi\1.2.16\jffi-1.2.16.jar;C:\Users\VssAdministrator\.m2\repository\com\github\jnr\jffi\1.2.16\jffi-1.2.16-native.jar;C:\Users\VssAdministrator\.m2\repository\org\jruby\joni\joni\2.1.15\joni-2.1.15.jar;C:\Users\VssAdministrator\.m2\repository\org\jruby\extras\bytelist\1.0.15\bytelist-1.0.15.jar;C:\Users\VssAdministrator\.m2\repository\org\jruby\jcodings\jcodings\1.0.27\jcodings-1.0.27.jar;C:\Users\VssAdministrator\.m2\repository\org\jruby\dirgra\0.3\dirgra-0.3.jar;C:\Users\VssAdministrator\.m2\repository\com\headius\invokebinder\1.10\invokebinder-1.10.jar;C:\Users\VssAdministrator\.m2\repository\com\headius\options\1.4\options-1.4.jar;C:\Users\VssAdministrator\.m2\repository\com\headius\unsafe-fences\1.0\unsafe-fences-1.0.jar;C:\Users\VssAdministrator\.m2\repository\com\jcraft\jzlib\1.1.3\jzlib-1.1.3.jar;C:\Users\VssAdministrator\.m2\repository\com\martiansoftware\nailgun-server\0.9.1\nailgun-server-0.9.1.jar;C:\Users\VssAdministrator\.m2\repository\joda-time\joda-time\2.8.2\joda-time-2.8.2.jar;C:\Users\VssAdministrator\.m2\repository\com\headius\modulator\1.0\modulator-1.0.jar;C:\Users\VssAdministrator\.m2\repository\org\jruby\jruby-stdlib\9.1.17.0\jruby-stdlib-9.1.17.0.jar org.jruby.Main -r ./siteconf20200413-3304-phwr7y.rb extconf.rb
[ERROR] 'java' is not recognized as an internal or external command,
[ERROR] operable program or batch file.
[ERROR] 
[ERROR] extconf failed, exit code 1
[ERROR] 
[ERROR] Gem files will remain installed in D:/a/1/s/lib/target/rubygems/gems/io-console-0.5.6 for inspection.
[ERROR] Results logged to D:/a/1/s/lib/target/rubygems/extensions/universal-java-1.8/2.3.0/io-console-0.5.6/gem_make.out
[ERROR] -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
[ERROR] 
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <goals> -rf :jruby-stdlib
```

Seems like some sort of path issue when building on Windows? Thoughts @mkristian?

See 52f30124e0b707100f008cf73ca4de5c816c10d1 for a revert of adding `io-console` to default gems.

**Comment by headius** [comment]
Found while attempting to update to latest IRB gem in #6172.

**Comment by mkristian** [comment]
hmm - it can not find `java`. maybe the assumption that java is on the path is wrong - at least it looks like.

**Comment by headius** [comment]
@mkristian Yeah that's what it looks like... it's assuming `java` is in path but it's not somehow?

**Comment by mkristian** [comment]
maven executes jruby which does this `gem install`, this one gem has an extension and tries to spawn a new 'jruby process' to install the noop extension.

it would be interesting to know of jruby itself - without the maven wrapper - can install this gem.

**Comment by headius** [comment]
@mkristian It can indeed, because the gem's `extconf.rb` emits a no-op `Makefile` when running on JRuby. So this must be specific to how the `gem install` command is being launched during the build.

**Comment by mkristian** [comment]
```
installer = Gem::Installer.new( a.file.to_pathname,
                                        :wrappers => true,
                                        :ignore_dependencies => true,
                                        :install_dir => ghome )
        def installer.ensure_required_ruby_version_met; end
        installer.install
```
that is the snippet from lib/pom.rb which installs the gem, this was executed by maven-polyglot. the build does show which jruby it uses for this snippet, something like
```
[INFO] --- polyglot-maven-plugin:0.3.1:execute (install_gems) @ jruby-stdlib ---
using jruby 9.1.17.0
install gems unless already installed
```

**Comment by mkristian** [comment]
@headius are you saying, yes, jruby can install this gem on windows but it fails via this lib/pom.rb script ?

**Comment by headius** [comment]
@mkristian I have not confirmed that, actually. Should be a matter of just running `gem install io-console` but I don't have a Windows system handy at the moment.

**Comment by mkristian** [comment]
but other question: what can I do when I execute JRuby to help it find the `java` executable ?

**Comment by headius** [comment]
@mkristian The jruby bash script does try to set `JAVACMD` if it is not already set, but I don't think the executable (used on Windows) does that. Naively, System.getProperty("java.home") + "/bin"` ought to be the right place, I guess.

My confusion here is why this `java` is not the full path to begin with. I think it's coming from RbConfig.ruby?

**Comment by headius** [comment]
@mkristian Aha, looks like we must be following this `classpath_launcher` logic when determining the command to use inside a Maven build:

https://github.com/jruby/jruby/blob/0af1b09fa7525e5216d4db737bd344f7996c074f/lib/ruby/stdlib/rubygems/defaults/jruby.rb#L11

That logic eventually leads here, where it attempts to find a `java` command using `java.home` property:

https://github.com/jruby/jruby/blob/4b749c3542ae36a8c3ca069bcbfce00871d612b8/core/src/main/java/org/jruby/util/ClasspathLauncher.java

I assume for whatever reason we are unable to resolve a `java` command on the filesystem and fall back on the `hope it's in path` section.

**Comment by mkristian** [comment]
it almost looks like 'java.home' is not set.

this scriplet is executed via an instance of IsolatedScriptingContainer 

I wonder if JAVA_HOME is set - as most windows installation do this as far I remember

**Comment by headius** [comment]
@mkristian Yeah it would seem to me very strange if `java.home` isn't set.

**Comment by headius** [comment]
Perhaps the path structure of JDK on Windows has changed?

**Comment by headius** [comment]
I figured out the problem while working on #6103.

The execution of the native extension build requires rbconfig. However rbconfig now starts its lifecycle on the filesystem, in lib/ruby/stdlib/rbconfig.rb.

When the gem install plugin attempts to run JRuby to install the gem, the extension build command line (`java -cp ...`) that it executes does not include a jruby.home setting:

```
/Library/Java/JavaVirtualMachines/adoptopenjdk-8.jdk/Contents/Home/jre/bin/java -cp ... org.jruby.Main -r ./siteconf20200617-83059-5knf5f.rb extconf.rb
```

I believe this command line is generated within RubyGems, and does not pick up any jruby.home from the outer plugin.

**Comment by headius** [comment]
Note also that this only affects non-SNAPSHOT builds, due to these lines:

https://github.com/jruby/jruby/blob/82bedae06cea4b3330661383b9fe1fbec0e8ca71/test/pom.rb#L120-L122

For now I'm just going to remove this, since release builds shouldn't ship the launcher anyway.

cc @mkristian @enebo Seems like we need a better solution here.

**Comment by mkristian** [comment]
@headius your are saying Windows fails when building a release. do we build release on windows anyways ? i.e. would it be OK to install this jruby-launcher only on non-windows machine on release version. just following the logic . . .

**Comment by headius** [comment]
@mkristian I'm saying all non-SNAPSHOT builds will fail this way. I was building 9.2 branch in the report above, which had not been updated to use a SNAPSHOT version.

It fails because the gem install tries to determine the jruby command line to use, which in an embedded scenario is the `java ...` command line. That does not provide a `jruby.home`, so `rbconfig` cannot be found. I believe this is broken on all platforms if you attempt to have the
gem-maven-plugin install the jruby-launcher gem.

**Comment by headius** [comment]
@mkristian I've worked around this for now by hardcoding to 9.2.9 (the last version to support --no-rdoc, for #6103) and removing the jruby launcher install altogether.

We need:

* a new gem-maven-plugin based on 9.2.11.1, not using `--no-rdoc` and `--no-ri`
* something in the install mojo that makes the RubyGems-created native extension build (#6103) command line to still find the appropriate jruby.home. This might be switching the mojo to use jruby-complete, setting some environment, or something else. (this bug)

My fixes are just workarounds, so this and #6103 remain open until the above can happen.

**Comment by headius** [comment]
See #6285 for the full set of commits leading up to and working around this and #6103.

**Comment by headius** [comment]
The io-console gem has been reworked to have a -java version that does not require an extension build, so that part is fixed. The remaining issue with rbconfig not being available in the bare jar file is documented in #6362.
